Adding ibm-dbt-db2 to the Trusted Adapters documentation (#9487)
fishtown-analytics/docs.getdbt.com: website/docs/docs/local/connect-data-platform/ibm-db2-setup.md

<SetUpPages meta={frontMatter.meta}/>

## Connecting to IBM Db2

To connect dbt with IBM Db2, you need to configure a profile in your `profiles.yml` file located in the `.dbt/` directory of your home folder. The following is an example configuration for connecting to IBM Db2 instances:

<File name='~/.dbt/profiles.yml'>

```yaml
my_db2_project:
  outputs:
    dev:
      type: db2
      host: [hostname]
      port: 50000
      database: [database name]
      schema: [schema name]
      username: [username]
      password: [password]
      threads: [1 or more]

  target: dev

```

</File>

### Connection with SSL/TLS

For secure connections, you can configure SSL/TLS parameters:

<File name='~/.dbt/profiles.yml'>

```yaml
my_db2_project:
  outputs:
    prod:
      type: db2
      host: [hostname]
      port: 50001
      database: [database name]
      schema: [schema name]
      username: [username]
      password: [password]
      threads: [1 or more]
      # SSL/TLS settings
      security: SSL
      ssl_server_certificate: /path/to/server-ca.crt
      ssl_client_hostname_validation: true
      retries: 3

  target: prod

```

</File>

### Connection using DSN

> Alternatively, you can use a DSN (Data Source Name) configured in your Db2 client:

Reviewer: ```suggestion
Alternatively, you can use a Data Source Name (DSN) configured in your Db2 client:
```

<File name='~/.dbt/profiles.yml'>

```yaml
my_db2_project:
  outputs:
    dev:
      type: db2
      dsn: MY_DB2_DSN
      schema: [schema name]
      threads: [1 or more]

  target: dev

```

</File>

## Host parameters

> The following profile fields are required to configure IBM Db2 connections.

Reviewer: ```suggestion
Use the following required and optional profile fields to configure IBM Db2 connections.
```

| `username`| Required | Username for authentication. | `db2user` |
| `password`| Required | Password for authentication. | `password` |
| `threads` | Optional | Number of threads for parallel execution. | `8` |
|   `dsn`   | Optional | Data Source Name (alternative to host/port/database). | `MY_DB2_DSN` |

*Not required if using `dsn`

### Optional SSL/TLS Parameters

| Option    | Required/Optional | Description | Example  |
| --------- | ------- | ------- | ----------- |
| `security` | Optional | Security protocol. Use `SSL` to enable SSL/TLS. | `SSL` |
| `ssl_server_certificate` | Optional | Path to server CA certificate file for SSL verification. | `/path/to/server-ca.crt` |
| `ssl_client_keystore` | Optional | Path to client keystore database (.kdb file) for mutual TLS. | `/path/to/client.kdb` |
| `ssl_client_keystash` | Optional | Path to client keystash file (.sth file). | `/path/to/client.sth` |
| `ssl_client_hostname_validation` | Optional | Enable hostname verification (true/false). | `true` |
| `retries` | Optional | Number of connection retry attempts. | `3` |

## Schemas and databases

> When selecting the database and the schema, make sure the user has read and write access to both. This selection does not limit your ability to query the database. Instead, they serve as the default location for where tables and views are materialized.

Reviewer: ```suggestion
When selecting the database and the schema, make sure you have read and write access to both. This selection does not limit your ability to query the database. Instead, they serve as the default location for where tables and views are materialized.
```

| dbt Type | Db2 Type |
|----------|----------|
| string | VARCHAR |
| text | VARCHAR(max_length) |
| integer | INTEGER |
| bigint | BIGINT |
| float | FLOAT |
| numeric | DECIMAL |
| boolean | BOOLEAN |
| timestamp | TIMESTAMP |
| date | DATE |
| time | TIME |

## Incremental models

The `ibm-dbt-db2` adapter supports two incremental strategies:

1. **merge** (default) - Uses Db2's MERGE statement for efficient upserts
2. **delete+insert** - Deletes matching records then inserts new ones

Example incremental model:

```sql
{{
  config(
    materialized='incremental',
    unique_key='id',
    incremental_strategy='merge'
  )
}}

SELECT * FROM source_table
{% if is_incremental() %}
WHERE updated_at > (SELECT MAX(updated_at) FROM {{ this }})
{% endif %}
```

## Case sensitivity

Db2 uppercases unquoted identifiers by default. The adapter handles this automatically, but be aware:

- Unquoted table/column names will be uppercased
- Use quotes in your SQL to preserve case: `"MyTable"` vs `MYTABLE`

## Notes

- The `ibm-dbt-db2` adapter is built on the `ibm_db` Python driver (version 3.2.8) which is automatically installed with the adapter.

> - **Python Version Requirements**: Requires Python 3.10, 3.11, or 3.12. Python 3.9 is not supported due to dbt-core 1.11+ dependency requirements. Python 3.13+ has not been tested yet.

Reviewer: ```suggestion
- **Python Version Requirements**: Requires Python 3.10, 3.11, or 3.12. Python 3.9 is not supported due to <Constant name="core" /> v1.11+ dependency requirements. Python 3.13+ has not been tested yet.
```
